File: week-16/bkryusim/공원_산책.py
def solution(park, routes):
    height = len(park)
    width = len(park[0])

    for r in range(height):
        for c in range(width):
            if park[r][c] == 'S':
                position = (r, c)

    move = {
        'E': (0, 1),
        'W': (0, -1),
        'N': (-1, 0),
        'S': (1, 0)
    }

    for route in routes:
        direction, distance = route.split()
        distance = int(distance)

        moved = (position[0] + move[direction][0] * distance, position[1] + move[direction][1] * distance)

        noop = False
        if moved[0] >= height or moved[0] < 0 or moved[1] < 0 or moved[1] >= width:
            continue

        for i in range(1, distance + 1):
            if park[position[0] + move[direction][0] * i][position[1] + move[direction][1] * i] == 'X':
                noop = True
                break
        if noop:
            continue

        position = (position[0] + move[direction][0] * distance, position[1] + move[direction][1] * distance)

    return list(position)

File: week-16/dahyen0o/공원-산책.java
import java.util.*;

class Solution {
    
    private static final Map<Character, int[]> dirs = Map.of(
        'N', new int[] {-1, 0},
        'S', new int[] {+1, 0},
        'W', new int[] {0, -1},
        'E', new int[] {0, +1}
    );
    
    private boolean[][] park;
    private int[] dog = new int[2];
    
    public int[] solution(String[] park, String[] routes) {
        this.park = new boolean[park.length][park[0].length()];
                
        for(int r = 0; r < park.length; r++) {
            for(int c = 0; c < park[r].length(); c++) {
                switch(park[r].charAt(c)) {
                    case 'S':
                        dog[0] = r;
                        dog[1] = c;
                        this.park[r][c] = true;
                        break;
                    case 'O':
                        this.park[r][c] = true;
                        break;
                }
            }
        }
        
        for(final String route : routes) {
            final String[] cmd = route.split(" ");
            move(cmd[0].charAt(0), Integer.parseInt(cmd[1]));
        }
        
        return dog;
    }
    
    private void move(final char DIR, final int DEPTH) {
        if(!canMove(DIR, DEPTH)) return;
        
        dog[0] += dirs.get(DIR)[0] * DEPTH;
        dog[1] += dirs.get(DIR)[1] * DEPTH;
    }
    
    private boolean canMove(final char DIR, final int DEPTH) {
        final int[] curr = new int[]{dog[0], dog[1]};
        final int[] dir = dirs.get(DIR);
        
        for(int depth = 0; depth < DEPTH; depth++) {
            curr[0] += dir[0];
            curr[1] += dir[1];
            if(!inBound(curr) || !park[curr[0]][curr[1]]) 
                return false;
        }
        return true;
    }
    
    private boolean inBound(final int[] curr) {
        return curr[0] >= 0 && curr[1] >= 0 && curr[0] < park.length && curr[1] < park[0].length;
    }
}
